File: netlify/functions/netlify_function.py
# Netlify/Vercel Serverless Function
# This handles Stripe webhook and generates Telegram invite link

import json
import os
import stripe
import asyncio
import threading
import aiohttp
import requests
from telegram import Bot
from datetime import datetime, timedelta

# Environment variables
stripe.api_key = os.environ.get('STRIPE_SECRET_KEY')
TELEGRAM_BOT_TOKEN = os.environ.get('TELEGRAM_BOT_TOKEN')
TELEGRAM_GROUP_ID = os.environ.get('TELEGRAM_GROUP_ID', '-1003798603747')
STRIPE_WEBHOOK_SECRET = os.environ.get('STRIPE_WEBHOOK_SECRET')

# Telegram Bot API base URL
TELEGRAM_API_URL = f'https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}'

# Initialize Telegram Bot for creating links
bot = Bot(token=TELEGRAM_BOT_TOKEN) if TELEGRAM_BOT_TOKEN else None

# Simple in-memory storage (use Redis or database in production)
invite_links = {}


def run_async(coro):
    """Run async code in serverless environment"""
    result = {'value': None, 'error': None}

    def run_in_thread():
        try:
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
            result['value'] = loop.run_until_complete(coro)
            loop.close()
        except Exception as e:
            result['error'] = e

    thread = threading.Thread(target=run_in_thread)
    thread.start()
    thread.join()

    if result['error']:
        raise result['error']
    return result['value']


async def create_telegram_invite_link():
    """Create one-time Telegram invite link"""
    expire_date = int((datetime.now() + timedelta(seconds=3600)).timestamp())
    invite_link = await bot.create_chat_invite_link(
        chat_id=TELEGRAM_GROUP_ID,
        expire_date=expire_date,
        member_limit=1
    )
    return invite_link.invite_link


async def unban_user(user_id):
    """Unban a user from the Telegram group using direct API call"""
    try:
        api_url = f'{TELEGRAM_API_URL}/unbanChatMember'

        payload = {
            'chat_id': int(TELEGRAM_GROUP_ID),
            'user_id': int(user_id),
            'only_if_banned': False
        }

        response = requests.post(api_url, json=payload, timeout=10)
        data = response.json()

        print(f"Telegram API Response: {data}")

        if data.get('ok'):
            return {'success': True, 'message': 'User unbanned successfully', 'details': data}
        error_description = data.get('description', 'Unknown error')
        print(f"Telegram API error: {error_description}")
        return {'success': False, 'message': f'Failed to unban: {error_description}', 'details': data}
    except Exception as e:
        error_msg = str(e)
        print(f"Exception in unban_user: {error_msg}")
        return {'success': False, 'message': f'Error: {error_msg}'}


def unban_user_sync(user_id):
    """Synchronous wrapper for unban operation"""
    try:
        api_url = f'{TELEGRAM_API_URL}/unbanChatMember'

        payload = {
            'chat_id': int(TELEGRAM_GROUP_ID),
            'user_id': int(user_id),
            'only_if_banned': False
        }

        print(f"Sending unban request for user {user_id} to chat {TELEGRAM_GROUP_ID}")
        response = requests.post(api_url, json=payload, timeout=10)
        data = response.json()

        print(f"Telegram API Response: {data}")

        if data.get('ok'):
            return {'success': True, 'message': 'User unbanned successfully'}
        error_description = data.get('description', 'Unknown error')
        print(f"Telegram API error: {error_description}")
        return {'success': False, 'message': f'Failed to unban: {error_description}'}
    except Exception as e:
        error_msg = str(e)
        print(f"Exception in unban_user_sync: {error_msg}")
        return {'success': False, 'message': f'Error: {error_msg}'}


def handler(event, context):
    """Serverless function handler"""

    # Handle CORS preflight
    if event.get('httpMethod') == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': {
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Methods': 'POST, GET, OPTIONS',
                'Access-Control-Allow-Headers': 'Content-Type, Stripe-Signature',
            },
            'body': ''
        }

    # Handle unban request
    if event.get('httpMethod') == 'POST':
        try:
            body = json.loads(event.get('body', '{}'))

            # Check if this is an unban request
            if body.get('action') == 'unban' or 'unban' in event.get('path', ''):
                user_id = body.get('user_id')

                if not user_id:
                    return {
                        'statusCode': 400,
                        'headers': {
                            'Access-Control-Allow-Origin': '*',
                            'Content-Type': 'application/json',
                        },
                        'body': json.dumps({'success': False, 'message': 'user_id is required'})
                    }

                # Call synchronous unban function
                result = unban_user_sync(int(user_id))

                return {
                    'statusCode': 200,
                    'headers': {
                        'Access-Control-Allow-Origin': '*',
                        'Content-Type': 'application/json',
                    },
                    'body': json.dumps(result)
                }
        except json.JSONDecodeError:
            pass
        except Exception as e:
            return {
                'statusCode': 500,
                'headers': {
                    'Access-Control-Allow-Origin': '*',
                    'Content-Type': 'application/json',
                },
                'body': json.dumps({'success': False, 'message': str(e)})
            }

    # Handle GET request for invite link retrieval
    if event.get('httpMethod') == 'GET':
        path = event.get('path', '')
        payment_id = path.split('/')[-1]

        if payment_id in invite_links:
            return {
                'statusCode': 200,
                'headers': {
                    'Access-Control-Allow-Origin': '*',
                    'Content-Type': 'application/json',
                },
                'body': json.dumps({
                    'invite_link': invite_links[payment_id],
                    'status': 'ready'
                })
            }
        return {
            'statusCode': 202,
            'headers': {
                'Access-Control-Allow-Origin': '*',
                'Content-Type': 'application/json',
            },
            'body': json.dumps({'status': 'processing'})
        }

    # Handle Stripe webhook (only if it has Stripe signature)
    stripe_signature = event.get('headers', {}).get('stripe-signature', '')
    if stripe_signature and event.get('httpMethod') == 'POST':
        try:
            payload = event.get('body', '')

            # Verify webhook signature
            webhook_event = stripe.Webhook.construct_event(
                payload, stripe_signature, STRIPE_WEBHOOK_SECRET
            )

            # Handle checkout.session.completed event
            if webhook_event['type'] == 'checkout.session.completed':
                session = webhook_event['data']['object']
                payment_id = session['payment_intent']

                # Generate Telegram invite link
                invite_link = run_async(create_telegram_invite_link())

                # Store it
                invite_links[payment_id] = invite_link

                print(f"Generated invite link for payment {payment_id}")

            return {
                'statusCode': 200,
                'body': json.dumps({'received': True})
            }

        except Exception as e:
            print(f"Stripe Error: {str(e)}")
            return {
                'statusCode': 400,
                'body': json.dumps({'error': str(e)})
            }

    # Handle health check
    if event.get('httpMethod') == 'GET' and 'health' in event.get('path', ''):
        return {
            'statusCode': 200,
            'headers': {
                'Access-Control-Allow-Origin': '*',
                'Content-Type': 'application/json',
            },
            'body': json.dumps({
                'status': 'healthy',
                'bot_token_present': bool(TELEGRAM_BOT_TOKEN),
                'group_id': TELEGRAM_GROUP_ID,
                'function': 'netlify_function'
            })
        }

    # Default response if no handler matched
    return {
        'statusCode': 400,
        'headers': {
            'Access-Control-Allow-Origin': '*',
            'Content-Type': 'application/json',
        },
        'body': json.dumps({'error': 'Invalid request'})
    }


# For Vercel

def main(request):
    return handler(request, None)

File: netlify/functions/netlify_function.js
// Netlify Serverless Function - Unban & Invite Link Generator
const https = require('https');

const TELEGRAM_BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const TELEGRAM_GROUP_ID = process.env.TELEGRAM_GROUP_ID || '-1003798603747';
const TELEGRAM_API_URL = `https://api.telegram.org/bot${TELEGRAM_BOT_TOKEN}`;

// Helper to make HTTPS POST requests
function telegramApiCall(method, payload) {
  return new Promise((resolve, reject) => {
    const data = JSON.stringify(payload);
    const url = new URL(`${TELEGRAM_API_URL}/${method}`);
    
    const options = {
      hostname: url.hostname,
      path: url.pathname,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(data),
      },
    };

    const req = https.request(options, (res) => {
      let body = '';
      res.on('data', (chunk) => { body += chunk; });
      res.on('end', () => {
        try {
          resolve(JSON.parse(body));
        } catch (e) {
          reject(new Error(`Failed to parse response: ${body}`));
        }
      });
    });

    req.on('error', reject);
    req.setTimeout(10000, () => {
      req.destroy();
      reject(new Error('Request timed out'));
    });
    req.write(data);
    req.end();
  });
}

// CORS headers
const corsHeaders = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'POST, GET, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type, Stripe-Signature',
  'Content-Type': 'application/json',
};

exports.handler = async (event, context) => {
  // Handle CORS preflight
  if (event.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers: corsHeaders,
      body: '',
    };
  }

  // Handle POST requests
  if (event.httpMethod === 'POST') {
    try {
      const body = JSON.parse(event.body || '{}');

      // Handle unban request
      if (body.action === 'unban') {
        const userId = body.user_id;

        if (!userId) {
          return {
            statusCode: 400,
            headers: corsHeaders,
            body: JSON.stringify({ success: false, message: 'user_id is required' }),
          };
        }

        console.log(`Sending unban request for user ${userId} to chat ${TELEGRAM_GROUP_ID}`);

        const result = await telegramApiCall('unbanChatMember', {
          chat_id: parseInt(TELEGRAM_GROUP_ID),
          user_id: parseInt(userId),
          only_if_banned: false,
        });

        console.log('Telegram API Response:', JSON.stringify(result));

        if (result.ok) {
          return {
            statusCode: 200,
            headers: corsHeaders,
            body: JSON.stringify({ success: true, message: 'User unbanned successfully' }),
          };
        }

        const errorDesc = result.description || 'Unknown error';
        console.log('Telegram API error:', errorDesc);
        return {
          statusCode: 200,
          headers: corsHeaders,
          body: JSON.stringify({ success: false, message: `Failed to unban: ${errorDesc}` }),
        };
      }

      // Handle create invite link request
      if (body.action === 'create_invite') {
        const expireDate = Math.floor(Date.now() / 1000) + 3600;
        const result = await telegramApiCall('createChatInviteLink', {
          chat_id: parseInt(TELEGRAM_GROUP_ID),
          expire_date: expireDate,
          member_limit: 1,
        });

        if (result.ok) {
          return {
            statusCode: 200,
            headers: corsHeaders,
            body: JSON.stringify({ success: true, invite_link: result.result.invite_link }),
          };
        }

        return {
          statusCode: 200,
          headers: corsHeaders,
          body: JSON.stringify({ success: false, message: result.description || 'Failed to create invite link' }),
        };
      }

      // Default POST response
      return {
        statusCode: 400,
        headers: corsHeaders,
        body: JSON.stringify({ success: false, message: 'Unknown action. Use action: "unban" or "create_invite"' }),
      };

    } catch (e) {
      console.error('Error:', e.message);
      return {
        statusCode: 500,
        headers: corsHeaders,
        body: JSON.stringify({ success: false, message: e.message }),
      };
    }
  }

  // Handle GET requests (health check)
  if (event.httpMethod === 'GET') {
    return {
      statusCode: 200,
      headers: corsHeaders,
      body: JSON.stringify({
        status: 'healthy',
        bot_token_present: !!TELEGRAM_BOT_TOKEN,
        group_id: TELEGRAM_GROUP_ID,
        function: 'netlify_function',
      }),
    };
  }

  // Default response
  return {
    statusCode: 400,
    headers: corsHeaders,
    body: JSON.stringify({ error: 'Invalid request method' }),
  };
};
